library/ssl_client.c

```diff
@@ -843,6 +887,39 @@ static int ssl_prepare_client_hello( mbedtls_ssl_context *ssl )
         }
     }
 
+#if defined(MBEDTLS_SSL_PROTO_TLS1_3) && \
+    defined(MBEDTLS_SSL_SERVER_NAME_INDICATION)
+    if( ssl->handshake->resume )
+    {
+        if( ssl->hostname != NULL && ssl->session_negotiate->hostname != NULL )
```

Comment:
```
if( ssl->hostname != NULL && ssl->session_negotiate->hostname != NULL )
{
    ....
}
else if( ssl->hostname != NULL || ssl->session_negotiate->hostname != NULL )
{
     return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
}
```
This will miss one case when both ssl->hostname and ssl->session_negotiate->hostname are NULL.

Reply:
If you initialize `hostname_mismatch` to 0 this can work and is indeed more compact.